### PR TITLE
BUG: Fix maybe_convert_numeric for unhashable objects

### LINKED ISSUE #13324: to_numeric(errors='coerce') broken when using unhashable objects
When using pandas.to_numeric with errors='coerce' and non hashable objects in the series/dataframe pandas returns the same series/dataframe as passed in, instead of coercing the non-numeric objects to nans.

```python
python -c "import pandas as pd; s = pd.Series([[13], 1.0, 'apple']); print(pd.to_numeric(s, errors='coerce'))"
```

The error is due to this line

https://github.com/pydata/pandas/blob/4de83d25d751d8ca102867b2d46a5547c01d7248/pandas/src/inference.pyx#L572

More generally this will affect any call to ``maybe_convert_numeric`` that has unhashable objects in the ``values`` param.

Suggested fix might be to update this check with something like

```python
if val.__hash__ is not None and val in na_values:
    blah
```
As there should never be a case where you would expect an unhashable object to be in a set of na_values

#### output of ``pd.show_versions()``
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.11.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-22-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_IE.UTF-8

pandas: 0.18.1
nose: 1.3.7
pip: 8.1.2
setuptools: 19.0
Cython: None
numpy: 1.11.0
scipy: None
statsmodels: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.5.3
pytz: 2016.4
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.8
boto: None
pandas_datareader: None

### PR DESCRIPTION
**Pull request by RogerThomas**
- [x] closes #13324
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/test_infer_and_convert.py [hunk 1]
```diff
@@ -102,6 +102,12 @@ def test_scientific_no_exponent(self):
         result = lib.maybe_convert_numeric(arr, set(), False, True)
         self.assertTrue(np.all(np.isnan(result)))
 
+    def test_convert_non_hashable(self):
+        arr = np.array([[10.0, 2], 1.0, 'apple'])
```
Comment (jreback) [inline pandas/tests/test_infer_and_convert.py]: add the issue number here as a comment

Comment (jreback) [inline pandas/tests/test_infer_and_convert.py]: test usage of pd.to_numeric as well (tests in pandas/tools

File: doc/source/whatsnew/v0.18.2.txt [hunk 2]
```diff
@@ -317,6 +317,7 @@ Bug Fixes
 - Bug in ``groupby`` where ``apply`` returns different result depending on whether first result is ``None`` or not (:issue:`12824`)
 
 
+- Bug in ``pd.to_numeric`` when ``errors='coerce'`` returns same dataframe/series when dataframe/series contains non-hashable objects (:issue:`13324`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: Bug in ``pd.to_numeric`` when ``errors='coerce'`` and input contains non-hashable objects....

**Comment by RogerThomas** [comment]
@jreback i've made the changes you suggested

File: pandas/tests/test_infer_and_convert.py [hunk 3]
```diff
@@ -102,6 +102,12 @@ def test_scientific_no_exponent(self):
         result = lib.maybe_convert_numeric(arr, set(), False, True)
         self.assertTrue(np.all(np.isnan(result)))
 
+    def test_convert_non_hashable(self):
+        # Test for Bug #13324
+        arr = np.array([[10.0, 2], 1.0, 'apple'])
+        result = lib.maybe_convert_numeric(arr, set(), False, True)
+        np.testing.assert_array_equal(result, np.array([np.nan, 1.0, np.nan]))
```
Comment (jreback) [inline pandas/tests/test_infer_and_convert.py]: use ``tm.assert_numpy_array_equal`` (we don't use th numpy testing functions at all; in fact there is a check again them; though this one slipped thru)

File: pandas/tools/tests/test_util.py [hunk 4]
```diff
@@ -279,6 +279,12 @@ def test_period(self):
         # res = pd.to_numeric(pd.Series(idx, name='xxx'))
         # tm.assert_series_equal(res, pd.Series(idx.asi8, name='xxx'))
 
+    def test_non_hashable(self):
+        # Test for Bug #13324
+        s = pd.Series([[10.0, 2], 1.0, 'apple'])
+        res = pd.to_numeric(s, errors='coerce')
```
Comment (jreback) [inline pandas/tools/tests/test_util.py]: run thru ignore, raise as well

**Comment by jreback** [comment]
looks good. ping when green.

**Comment by jreback** [comment]
@RogerThomas thanks!

I was just experimenting with something and:

so maybe should change this to ``if getattr(val, '__hash__', None) is not None``

though can't seem to get this to trigger

```
In [1]: 'a'.__hash__
Out[1]: <method-wrapper '__hash__' of str object at 0x1002e1508>

In [2]: None.__hash__
Out[2]: <method-wrapper '__hash__' of NoneType object at 0x100180c40>

In [3]: 1.__hash__
  File "<ipython-input-3-e97b483155ff>", line 1
    1.__hash__
             ^
SyntaxError: invalid syntax
```

**Comment by jreback** [comment]
if you want to see if you can generate a test that makes this fail (not sure). If so , can you do a followup?

**Comment by RogerThomas** [comment]
Hey @jreback, I think this is just due to python's syntax parser.
```python
In [1]: a = 1

In [2]: a.__hash__
Out[2]: <method-wrapper '__hash__' of int object at 0xa63b00>
```
I don't think this could ever fail

**Comment by jreback** [comment]
oh ok that makes sense thanks

**Pull request closed without merging**